File: SendRecv/deadlock/deadlock.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<mpi.h>

int main(int argc,char *argv[])
{
    int myid, ntasks, n, m, i,j, count;
    const int maxN=1000000;
    double *outbuf, *inbuf;
    MPI_Status status;
    
    MPI_Init(&argc,&argv);
    
    MPI_Comm_size(MPI_COMM_WORLD,&ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD,&myid);
    outbuf=malloc(sizeof(double)*maxN);
    inbuf=malloc(sizeof(double)*maxN);
    
    for (n=1; n<=maxN;n*=10){
        if (myid == 0){
            for(i=0; i<n; i++) outbuf[i]=0;
            for(j=1;j<ntasks;j++){
                MPI_Send(outbuf,n,MPI_DOUBLE,j,0,MPI_COMM_WORLD);
                MPI_Recv(inbuf,n,MPI_DOUBLE,j,1,MPI_COMM_WORLD,&status);
                MPI_Get_count(&status,MPI_DOUBLE,&count);
                printf("0 received %d numbers\n",count);
            }
        }
        else {
            for(i=0; i<n; i++) outbuf[i]=1;
            MPI_Send(outbuf,n,MPI_DOUBLE,0,1,MPI_COMM_WORLD);
            MPI_Recv(inbuf,n,MPI_DOUBLE,0,0,MPI_COMM_WORLD,&status);      
            MPI_Get_count(&status,MPI_DOUBLE,&count);
            printf("%d received %d numbers\n",myid,count);
        }     
    }
    free(outbuf);
    free(inbuf);
    MPI_Finalize();
}

File: SendRecv/deadlock/deadlock.f90

PROGRAM deadlock
  USE mpi
  IMPLICIT NONE
  INTEGER,PARAMETER::size=10000000
  INTEGER :: err, myid,ntasks,i, n, m, count, dtype, stat(mpi_status_size)
  DOUBLE PRECISION :: outbuf(size),inbuf(size)

  CALL mpi_init(err)
  CALL mpi_comm_rank(mpi_comm_world,myid,err)
  CALL mpi_comm_size(mpi_comm_world,ntasks,err)


  DO m=1,7
    
    n=10**m

    IF ( myid == 0 ) THEN
      outbuf=0
      DO i=1,ntasks-1
         CALL mpi_send(outbuf,n,mpi_double_precision,i,0,mpi_comm_world,err)
         CALL mpi_recv(inbuf,n,mpi_double_precision,i,1,mpi_comm_world,stat,err)
         CALL mpi_get_count(stat,mpi_double_precision,count,err)
         PRINT *,'0 received',count,'numbers'
      end DO
    ELSE
      outbuf=1
      CALL mpi_send(outbuf,n,mpi_double_precision,0,1,mpi_comm_world,err)
      CALL mpi_recv(inbuf,n,mpi_double_precision,0,0,mpi_comm_world,stat,err)
      CALL mpi_get_count(stat,mpi_double_precision,count,err)
      PRINT *,myid,' received',count,'numbers'
    END IF

  END DO

  CALL mpi_finalize(err)

END PROGRAM deadlock
